gateway/health_check.go

```diff
@@ -0,0 +1,182 @@
+package gateway
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/TykTechnologies/tyk/config"
+	"github.com/TykTechnologies/tyk/storage"
+	"github.com/sirupsen/logrus"
+)
+
+type (
+	HealthCheckStatus string
+
+	HealthCheckComponentType string
+)
+
+const (
+	Pass HealthCheckStatus = "pass"
+	Fail                   = "fail"
+	Warn                   = "warn"
+
+	Component HealthCheckComponentType = "component"
+	Datastore                          = "datastore"
+	System                             = "system"
+)
+
+var (
+	healthCheckInfo atomic.Value
+	healthCheckLock sync.Mutex
+)
+
+func setCurrentHealthCheckInfo(h map[string]HealthCheckItem) {
+	healthCheckLock.Lock()
+	healthCheckInfo.Store(h)
+	healthCheckLock.Unlock()
+}
+
+func getHealthCheckInfo() map[string]HealthCheckItem {
+	healthCheckLock.Lock()
+	ret := healthCheckInfo.Load().(map[string]HealthCheckItem)
+	healthCheckLock.Unlock()
+	return ret
+}
+
+type HealthCheckResponse struct {
+	Status      HealthCheckStatus          `json:"status"`
+	Version     string                     `json:"version,omitempty"`
+	Output      string                     `json:"output,omitempty"`
+	Description string                     `json:"description,omitempty"`
+	Details     map[string]HealthCheckItem `json:"details,omitempty"`
+}
+
+type HealthCheckItem struct {
+	Status        HealthCheckStatus `json:"status"`
+	Output        string            `json:"output,omitempty"`
+	ComponentType string            `json:"componentType,omitempty"`
+	ComponentID   string            `json:"componentId,omitempty"`
+	Time          string            `json:"time"`
+}
+
+func initHealthCheck(ctx context.Context) {
+	if config.Global().LivenessCheck.Enabled {
+
+		setCurrentHealthCheckInfo(make(map[string]HealthCheckItem, 3))
+
+		go func(ctx context.Context) {
+			ticker := time.NewTicker(time.Second * 10)
+
+			for {
+				select {
+				case <-ctx.Done():
+
+					ticker.Stop()
+					mainLog.WithFields(logrus.Fields{
+						"prefix": "health-check",
+					}).Debug("Stopping Health checks for all components")
+					return
+
+				case <-ticker.C:
+					gatherHealthChecks()
+				}
+			}
+		}(ctx)
+	}
+}
+
+func gatherHealthChecks() {
```

Comment:
I would be careful here since in case of network issue you can get a timeout, and health check endpoint will take too long to respond. Instead, I would suggest running all three check-in 3 separate goroutines, like every second, and health check API will just depend on this data, and respond immediately.

Reply:
That's how it is implemented, API responds on the status, `gatherHealthChecks` runs periodically in another goroutine